botocore/regions.py

```diff
@@ -10,89 +10,194 @@
+    """Resolves regions and endpoints. Must be subclassed."""
+    def construct_endpoint(self, service_name, region_name):
+        """Resolves an endpoint for a service and region combination.
+
+        :type service_name: string
+        :param service_name: Name of the service to resolve an endpoint for
+            (e.g., s3)
+
+        :type region_name: string
+        :param region_name: Region/endpoint name to resolve (e.g., us-east-1)
+
+        :rtype: dict
+        :return: Returns a dict containing the following keys:
+            - partition: (string, required) Resolved partition name
+            - endpointName: (string, required) Resolved endpoint name
+            - hostname: (string, required) Hostname to use for this endpoint
+            - sslCommonName: (string) sslCommonName to use for this endpoint.
+            - credentialScope: (dict) Signature version 4 credential scope
+              - region: (string) region name override when signing.
+              - service: (string) service name override when signing.
+            - signatureVersions: (list<string>) A list of possible signature
+              versions, including s3, v4, v2, and s3v4
+            - protocols: (list<string>) A list of supported protocols
+              (e.g., http, https)
+            - ...: Other keys may be included as well based on the metadata
+        """
+        raise NotImplementedError
+
+    def get_available_partitions(self):
+        """Lists the partitions available to the endpoint resolver.
+
+        :return: Returns a list of partition names (e.g., ["aws", "aws-cn"]).
+        """
+        raise NotImplementedError
+
+    def get_available_endpoints(self, service_name, partition_name='aws',
+                                allow_non_regional=False):
+        """Lists the endpoint names of a particular partition.
+
+        :type service_name: string
+        :param service_name: Name of a service to list endpoint for (e.g., s3)
+
+        :type partition_name: string
+        :param partition_name: Name of the partition to limit endpoints to.
+            (e.g., aws for the public AWS endpoints, aws-cn for AWS China
+            endpoints, aws-us-gov for AWS GovCloud (US) Endpoints, etc.
 
-class EndpointResolver(object):
-    _CONSTRAINT_FUNCS = {
-        'startsWith': lambda x, y: str(x).startswith(y),
-        'notStartsWith': lambda x, y: not str(x).startswith(y),
-        'equals': lambda x, y: x == y,
-        'notEquals': lambda x, y: x != y,
-        'oneOf': lambda x, y: x in y
-    }
-    DEFAULT_SCHEME = 'https'
+        :type allow_non_regional: bool
+        :param allow_non_regional: Set to True to include endpoints that are
+             not regional endpoints (e.g., s3-external-1,
+             fips-us-gov-west-1, etc).
+        :return: Returns a list of endpoint names (e.g., ["us-east-1"]).
+        """
+        raise NotImplementedError
+
+
+class S3CompatResolver(BaseEndpointResolver):
```

Comment:
I wonder if it would be better to change this around a bit. It seems sort of weird that there is a S3CompatResolver that wraps the EndpointResolver (which follows the specification to a T) because if there are ever further backwards compatibilities that we need to add the pattern seems to be that we will keep wrapping the original EndpointResolver. So like:
```py
        def create_default_resolver():
            loader = self.get_component('data_loader')
            endpoints = loader.load_data('endpoints', use_ordered_dict=False)
            return OtherCompatResolver(S3CompatResolver(EndpointResolver(endpoints)))
```
Especially since there are already compatibility logic in both the endpoint and the client layers, I feel that some compatibility layer that conjoins all of the various compatibility logic in each class would be very convenient as it will clean up the codebase by removing the class and make it easier to find and wrap your head around all of the compatibility logic.